### PR TITLE
feat(renderer): set git icons in signcolumn

### LINKED ISSUE #553: Display git status in signcolumn
As I understand things right now, you leverage the signcolumn to display the lsp diagnostic status for files with potential issues.
While I like the idea of that, I'd rather use the signcolumn to display the git status information (thereby also preventing things like #72). Is that something you have thought about or would be willing to implement?

An additional cherry on top would be to have a configuraiton option whether git status signs are placed:
* to the right of file/folder names
* to the left of file/folder names
* in the sign column

Maybe you could leverage the same implementation to make the lsp diagnostic symbol location configurable in the same way? :thinking:

### PR DESCRIPTION
**Pull request by kyazdani42**
This is a work in progress. It works as is but there are some caveat that i'd like to discuss first.

Right, now, we will display the first icon returned by the list of icons from the node. Should we have a smarter way of doing that ?
Also, if displaying only one icons in signs, should we also always display one icon (after and before filename for instance ?)

TODO:
- [x] documentation

Fixes #553

### CODE REVIEW

File: lua/nvim-tree/diagnostics.lua [hunk 1]
```diff
@@ -32,7 +32,7 @@ local function add_sign(linenr, severity)
     return
   end
   local sign_name = sign_names[severity][1]
-  vim.fn.sign_place(1, GROUP, sign_name, buf, { lnum = linenr + 1 })
+  vim.fn.sign_place(1, GROUP, sign_name, buf, { lnum = linenr + 1, priority = 2 })
```
Comment (kyazdani42) [inline lua/nvim-tree/diagnostics.lua:35]: this will be higher priority than git signs

File: lua/nvim-tree/renderer/components/git.lua [hunk 2]
```diff
@@ -124,6 +126,17 @@ local git_hl = {
   [" A"] = "none",
 }
 
+function M.setup_signs()
+  local i = M.icon_state.icons.git_icons
```
Comment (kyazdani42) [inline lua/nvim-tree/renderer/components/git.lua:130]: this might not be initialized yet

File: lua/nvim-tree/renderer/init.lua [hunk 3]
```diff
@@ -17,11 +17,14 @@ local M = {
...
 local namespace_id = api.nvim_create_namespace "NvimTreeHighlights"
 
-local function _draw(bufnr, lines, hl)
+local function _draw(bufnr, lines, hl, signs)
   api.nvim_buf_set_option(bufnr, "modifiable", true)
   api.nvim_buf_set_lines(bufnr, 0, -1, false, lines)
   M.render_hl(bufnr, hl)
   api.nvim_buf_set_option(bufnr, "modifiable", false)
+  for _, sign in pairs(signs) do
+    vim.fn.sign_place(1, git.SIGN_GROUP, sign.sign, bufnr, { lnum = sign.lnum, priority = 1 })
```
Comment (kyazdani42) [inline lua/nvim-tree/renderer/init.lua:26]: not sure about the ID, i do not understand what it's used for. I've set it just like the one in diagnostics.

File: lua/nvim-tree/renderer/init.lua [hunk 4]
```diff
@@ -111,6 +116,7 @@ function M.setup(opts)
   }
 
   _padding.setup(opts)
+  git.setup_signs()
```
Comment (kyazdani42) [inline lua/nvim-tree/renderer/init.lua:119]: should we clear/setup signs on each draw ?

File: lua/nvim-tree/diagnostics.lua (hunk 1 above)
Comment (alex-courtis) [reply lua/nvim-tree/diagnostics.lua:35]: Tested OK.

We should document that behaviour.

File: lua/nvim-tree/renderer/init.lua (hunk 4 above)
Comment (alex-courtis) [reply lua/nvim-tree/renderer/init.lua:119]: It is working like this. Changing highlight groups immediately updates.

File: lua/nvim-tree/renderer/components/git.lua (hunk 2 above)
Comment (alex-courtis) [reply lua/nvim-tree/renderer/components/git.lua:130]: It should be.
`M.icon_state = _icons.get_config()` will always be called before `setup_signs`, as renderer init requires git.

File: lua/nvim-tree/renderer/init.lua (hunk 3 above)
Comment (alex-courtis) [reply lua/nvim-tree/renderer/init.lua:26]: sign-identifier seems to be used for mark like functionality, which we don't really need.

You could just call sign_place with 0 and have vim allocate a unique one.

**Review by alex-courtis — APPROVED** [review]
Tested OK:
- diagnostic fail: git_placement: after, before, signcolumn
- external change
- change some highlight groups

**Review by alex-courtis — APPROVED** [review]
re-tested base case - OK

File: lua/nvim-tree/renderer/components/git.lua (hunk 2 above)
Comment (gegoune) [reply lua/nvim-tree/renderer/components/git.lua:130]: I have got
```lua
vim.g.nvim_tree_icons = {
  git = {
    unstaged = 'M',
    staged = 'S',
    unmerged = 'U',
    renamed = 'R',
    untracked = '?',
    deleted = 'D',
    ignored = 'I',
  },
}
```
but icons in `signcolumn` are (I guess) some default (x, ★). Should I open an issue for it?

Comment (kyazdani42) [reply lua/nvim-tree/renderer/components/git.lua:130]: are you properly setting up the nvim_tree_icons before requiring nvim-tree ?

Comment (gegoune) [reply lua/nvim-tree/renderer/components/git.lua:130]: Ah, thanks! I was setting those `g:` before `require('nvim-tree').setup()` until recently adding a pcall-require-guard to all my plugins. User error - all works fine now, thanks!

**Pull request merged**